include/pybind11/detail/internals.h

```diff
@@ -311,12 +311,27 @@ struct type_info {
 #endif
 
 /// On Linux/OSX, changes in __GXX_ABI_VERSION__ indicate ABI incompatibility.
-/// On MSVC, changes in _MSC_VER may indicate ABI incompatibility (#2898).
+/// On MSVC, mixing /MT and /MD will result in crashes. See (#4953)
+/// There is no macro for major version for MSVC, so we check for major version
+/// 19, 20, 21, 22 for now as major version 19 is MSVC 2015-2022 and we future
+/// proof for 3 major versions in the future
```

Comment:
I think it's better not to do that (future proof for 3 major versions), because the future is unpredictable. It's only important that we don't silently fall through to `""` like before. (You're doing that already.)

(I want to preserve some of the changes under my experimental #5411. Might take me a day or two to get back to that.)

Reply:
I wouldn't characterize it as unpredictable. The version has a predictable MMmm pattern and since there's no way to divide an integer and assign to another macro in the preprocessor, this is necessary.

Reply:
What I meant: It is likely but not certain that future MSVC versions will be ABI compatible if `(_MSC_VER) / 100` are identical.

I think it's better to adjust/review here when those versions are released and we start testing. (We can backport to older pybind11 versions as needed to make them work. That's safer than having them assume ABI compatibility that may not exist.)

Reply:
I don't understand the sentiment though. On other platforms like clang on macos, we assume that things work fine, but on certain platforms like msvc, we assume the worst.

Reply:
I don't think that's true.

We don't have anything like `version / 100` anywhere else. (Please correct me if I'm missing something.)

Unless ABI compatibility for `version / 100` is a documented feature for a given platform, I believe it's prudent not to jump to conclusions. Safer is better here in my judgement.

In the back of my mind:

* Obviously the "it will be ok" fall-through has proven to be harmful. Let's do less of that in the future.

* See #5296 for where my interest in more strictly enforced correctness is coming from. The conduit feature could work between any pair of bindings systems, but it has to be safe to not get a bad name.